src/utils/resource_metrics/resource_metrics.c

```diff
@@ -0,0 +1,240 @@
+                 (void *)resource_metrics_compare);
+}
+
+static int insert_resource(resource_metrics_set_t *set, label_set_t resource) {
+  resource_metrics_t *rm =
+      realloc(set->ptr, (set->num + 1) * sizeof(*set->ptr));
+  if (rm == NULL) {
+    return ENOMEM;
+  }
+  set->ptr = rm;
+
+  rm = set->ptr + set->num;
+  memset(rm, 0, sizeof(*rm));
+
+  int status = label_set_clone(&rm->resource, resource);
+  if (status != 0) {
+    return ENOMEM;
+  }
+  set->num++;
+
+  qsort(set->ptr, set->num, sizeof(*set->ptr),
+        (void *)resource_metrics_compare);
+  return 0;
+}
+
+static resource_metrics_t *
+lookup_or_insert_resource(resource_metrics_set_t *set, label_set_t resource) {
+  resource_metrics_t *ret = lookup_resource(set, resource);
+  if (ret != NULL) {
+    return ret;
+  }
+
+  int status = insert_resource(set, resource);
+  if (status != 0) {
+    ERROR("resource_metrics: insert_resource failed: %s", STRERROR(status));
+    return NULL;
+  }
+
+  ret = lookup_resource(set, resource);
+  assert(ret != NULL);
+  return ret;
+}
+
+static int compare_family_by_name(metric_family_t **a, metric_family_t **b) {
+  return strcmp((*a)->name, (*b)->name);
+}
+
+static metric_family_t *lookup_family(resource_metrics_t *rm,
+                                      metric_family_t const *fam) {
+  metric_family_t **ret =
+      bsearch(&fam, rm->families, rm->families_num, sizeof(*rm->families),
+              (void *)compare_family_by_name);
+  if (ret == NULL) {
+    return NULL;
+  }
+  return *ret;
+}
+
+static int insert_family(resource_metrics_t *rm, metric_family_t const *fam) {
+  metric_family_t **tmp =
+      realloc(rm->families, (rm->families_num + 1) * sizeof(*rm->families));
+  if (tmp == NULL) {
+    return ENOMEM;
+  }
+  rm->families = tmp;
+
+  /* NOTE: metric_family_clone also copies the resource attributes. This is not
+   * strictly required, since we have these attributes in rm. We keep the copies
+   * for now for the sake of simplicity. If memory consumption is a problem,
+   * this could be de-duplicated, at the cost of more complicated memory
+   * management. */
+  rm->families[rm->families_num] = metric_family_clone(fam);
+  if (rm->families[rm->families_num] == NULL) {
+    return errno;
+  }
+
+  metric_family_metric_reset(rm->families[rm->families_num]);
+  label_set_reset(&rm->families[rm->families_num]->resource);
+
+  rm->families_num++;
+
+  qsort(rm->families, rm->families_num, sizeof(*rm->families),
+        (void *)compare_family_by_name);
+  return 0;
+}
+
+static metric_family_t *lookup_or_insert_family(resource_metrics_t *rm,
+                                                metric_family_t const *fam) {
+  metric_family_t *ret = lookup_family(rm, fam);
+  if (ret != NULL) {
+    return ret;
+  }
+
+  int status = insert_family(rm, fam);
+  if (status != 0) {
+    ERROR("resource_metrics: insert_family failed: %s", STRERROR(status));
+    return NULL;
+  }
+
+  ret = lookup_family(rm, fam);
```

Comment:
Maybe insert could return the inserted pointer, to avoid need for 2nd lookup?

Reply:
Heh, this is very tempting ;) In the interest of not getting nerd-sniped at 23:30 I'm going to stick with "maybe later". :wink: